ethdb/privateapi/engine_test.go

```diff
@@ -0,0 +1,305 @@
+	tx, _ := db.BeginRw(ctx)
+	rawdb.WriteHeadBlockHash(tx, payload1Hash)
+	rawdb.WriteHeaderNumber(tx, payload1Hash, 100)
+	_ = tx.Commit()
+	// Now we try to sync the next payload again
+	go func() {
+		reply, err = backend.EngineExecutePayloadV1(ctx, mockPayload2)
+		done <- true
+	}()
+
+	<-done
+
+	require.NoError(err)
+	require.Equal(reply.Status, Syncing)
+	replyHash = gointerfaces.ConvertH256ToHash(reply.LatestValidHash)
+	require.Equal(replyHash[:], startingHeadHash[:])
+}
+
+func TestMockValidExecution(t *testing.T) {
+	db := memdb.New()
+	ctx := context.Background()
+	require := require.New(t)
+
+	makeTestDb(ctx, db)
+
+	reverseDownloadCh := make(chan types.Block)
+	statusCh := make(chan ExecutionStatus)
+	waitingForHeaders := true
+
+	backend := NewEthBackendServer(ctx, nil, db, nil, nil, &params.ChainConfig{TerminalTotalDifficulty: common.Big1}, reverseDownloadCh, statusCh, &waitingForHeaders)
+
+	var err error
+	var reply *remote.EngineExecutePayloadReply
+	done := make(chan bool)
+
+	go func() {
+		reply, err = backend.EngineExecutePayloadV1(ctx, mockPayload3)
+		done <- true
+	}()
+
+	<-reverseDownloadCh
+
+	statusCh <- ExecutionStatus{
+		HeadHash: payload3Hash,
+		Status:   Valid,
+	}
+	<-done
+
+	require.NoError(err)
+	require.Equal(reply.Status, Valid)
+	replyHash := gointerfaces.ConvertH256ToHash(reply.LatestValidHash)
+	require.Equal(replyHash[:], payload3Hash[:])
+}
+
+func TestMockInvalidExecution(t *testing.T) {
+	db := memdb.New()
+	ctx := context.Background()
+	require := require.New(t)
+
+	makeTestDb(ctx, db)
+
+	reverseDownloadCh := make(chan types.Block)
+	statusCh := make(chan ExecutionStatus)
+
+	waitingForHeaders := true
+	backend := NewEthBackendServer(ctx, nil, db, nil, nil, &params.ChainConfig{TerminalTotalDifficulty: common.Big1}, reverseDownloadCh, statusCh, &waitingForHeaders)
+
+	var err error
+	var reply *remote.EngineExecutePayloadReply
+	done := make(chan bool)
+
+	go func() {
+		reply, err = backend.EngineExecutePayloadV1(ctx, mockPayload3)
+		done <- true
+	}()
+
+	<-reverseDownloadCh
+
+	statusCh <- ExecutionStatus{
+		HeadHash: startingHeadHash,
+		Status:   Invalid,
+	}
+	<-done
+
+	require.NoError(err)
+	require.Equal(reply.Status, Invalid)
+	replyHash := gointerfaces.ConvertH256ToHash(reply.LatestValidHash)
+	require.Equal(replyHash[:], startingHeadHash[:])
+}
+
+func TestInvalidRequest(t *testing.T) {
+	db := memdb.New()
+	ctx := context.Background()
+	require := require.New(t)
+
+	makeTestDb(ctx, db)
+
+	reverseDownloadCh := make(chan types.Block)
+	statusCh := make(chan ExecutionStatus)
+	waitingForHeaders := true
+
+	backend := NewEthBackendServer(ctx, nil, db, nil, nil, &params.ChainConfig{TerminalTotalDifficulty: common.Big1}, reverseDownloadCh, statusCh, &waitingForHeaders)
+
+	var err error
+
+	done := make(chan bool)
+
+	go func() {
+		_, err = backend.EngineExecutePayloadV1(ctx, &types2.ExecutionPayload{
```

Comment:
Why is this payload invalid? It's not immediately clear. Please add a clarifying comment.

Reply:
Ok, so we are giving status invalid, this is meant to be done in stage execution, if that block gets executed with no error we give off valid if not invalid.

Reply:
Sorry, I'm not following. Is there some special logic somewhere to return invalid status for this test, or the payload specified in this test is somehow invalid and will be rejected due to a standard validity check?

Reply:
comment added

Reply:
In other words, why are we expecting "invalid" status?

Reply:
there is no special logic, Invalid is returned in case we have a problem executing the block for some reason. so invalid is returned if the block is a bad block

Reply:
in the test we mock the behaviour that the execute stage will have towards an invalid payload

Reply:
we are expecting invalid status because that's what the statusCh sent us, that will be supposed to happen in stage execution

Reply:
Sorry, I'm still not following. Where is the piece of code that sends invalid status to `statusCh` for `TestInvalidRequest`? Does it do because it was mocked to always return invalid status or because the payload in `TestInvalidRequest` is somehow invalid?

Reply:
it was mocked, `statusCh <- ExecutionStatus{Status: Invalid}`is the mock

Reply:
statusCh is supposed to be updated in stage execution where we can determine wheter the block is good or bad

Reply:
the payload itself is not invalid in itself

Reply:
I see `statusCh <- ExecutionStatus{Status: Invalid}` in `TestMockInvalidExecution`, but not in `TestInvalidRequest`. So my question stands.

Reply:
TestInvalidRequest is if we give a request that is not valid, like a malformed block. for example if it misses some fields, since they are all defined as pointers in gointerface

Reply:
All right.  So please put a comment into `TestInvalidRequest` clarifying that the request is invalid because field XXX is missing.

Reply:
i made it more clear now, sorry